### PR TITLE
Add blog post on building a Spring REST API and resource server

### PR DESCRIPTION
**Pull request by mraible**
@bdemers Can you please review? Also, I'm interested to know how difficult it'd be to refactor this post/example to use our Starter v1.0.

### CODE REVIEW

File: _source/_posts/2018-12-18-secure-spring-rest-api.md [hunk 1]
```diff
@@ -0,0 +1,283 @@
...
+    <artifactId>spring-boot-starter-security</artifactId>
+</dependency>
+<dependency>
+    <groupId>org.springframework.cloud</groupId>
+    <artifactId>spring-cloud-starter-oauth2</artifactId>
+</dependency>
+<dependency>
+    <groupId>com.okta.spring</groupId>
+    <artifactId>okta-spring-boot-starter</artifactId>
+    <version>0.6.1</version>
```
Comment (bdemers) [inline _source/_posts/2018-12-18-secure-spring-rest-api.md]: `1.0.0`

File: _source/_posts/2018-12-18-secure-spring-rest-api.md [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+
+> **NOTE:** If you are struggling, you can check the modifications in Git branch `step-1-security-dependencies`.
+
+For that, you need to modify `application.properties` as follows (use _client_id_ and _client_secret_ provided by Okta dashboard to your application):
+
+```properties
+okta.oauth2.issuer=https://{yourOktaDomain}/oauth2/default
+okta.oauth2.clientId=${clientId}
+okta.oauth2.clientSecret=${clientSecret}
+okta.oauth2.scopes=openid
```
Comment (bdemers) [inline _source/_posts/2018-12-18-secure-spring-rest-api.md]: You can probably use use:
```
okta.oauth2.issuer=https://{yourOktaDomain}/oauth2/default
okta.oauth2.client-id={clientId}
```

File: _source/_posts/2018-12-18-secure-spring-rest-api.md [hunk 3]
```diff
@@ -0,0 +1,283 @@
...
+
+```java
+package net.dovale.okta.secure_rest_api;
+
+import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity;
+import org.springframework.security.oauth2.config.annotation.web.configuration.EnableResourceServer;
+
+@EnableWebSecurity
+@EnableResourceServer
+public class SecurityConfig  {}
```
Comment (bdemers) [inline _source/_posts/2018-12-18-secure-spring-rest-api.md]: With Boot 2.1 and our new lib you should be able to do something like:
https://github.com/okta/samples-java-spring/blob/spring-boot-update/resource-server/src/main/java/com/okta/spring/example/ResourceServerExampleApplication.java#L21-L35

File: _source/_posts/2018-12-18-secure-spring-rest-api.md [hunk 4]
```diff
@@ -0,0 +1,283 @@
...
+
+The new `@EnableGlobalMethodSecurity(prePostEnabled = true)` annotation tells Spring to use [AOP](https://en.wikipedia.org/wiki/Aspect-oriented_programming)-like method security and `prePostEnabled = true` will enable _pre_ and _post_ annotations. Those annotations will enable us to define security programmatically for each endpoint.
+
+Now, make changes to `HelloWorldController.java` to create a _scope-protected_ endpoint:
+
+```java
+import org.springframework.security.access.prepost.PreAuthorize;
+import java.security.Principal;
+...
+@PreAuthorize("#oauth2.hasScope('profile')")
```
Comment (bdemers) [inline _source/_posts/2018-12-18-secure-spring-rest-api.md]: this is now `@PreAuthorize("hasAuthority('SCOPE_profile')")`

File: _source/_posts/2018-12-18-secure-spring-rest-api.md (hunk 1 above)
Comment (mraible) [reply _source/_posts/2018-12-18-secure-spring-rest-api.md]: If you use the 1.0.0 version of our starter, I don’t think this is necessary.

Comment (mraible) [reply _source/_posts/2018-12-18-secure-spring-rest-api.md]: @bdemers Are Spring Boot’s OAuth starters needed for login or resource server?

File: _source/_posts/2018-12-18-secure-spring-rest-api.md (hunk 3 above)
Comment (mraible) [reply _source/_posts/2018-12-18-secure-spring-rest-api.md]: @bdemers I tried the following:

```java
@EnableWebSecurity
@EnableGlobalMethodSecurity(prePostEnabled = true)
public class SecurityConfig extends WebSecurityConfigurerAdapter {

    protected void configure(HttpSecurity http) throws Exception {
        http.oauth2ResourceServer().jwt();
    }
}
```

But this doesn't seem to protect any controllers from being accessed. I also tried adding the following dependency, but it doesn't seem to make a difference.

		<dependency>
			<groupId>org.springframework.security</groupId>
			<artifactId>spring-security-oauth2-resource-server</artifactId>
		</dependency>

**Comment by mraible** [comment]
@bdemers In the interest of time, I'm going to publish this post "as is". When I figure out how to update it to use the 1.0 version of our starter, I'll update it and add a changelog. Can you please review and approve?

File: _source/_posts/2018-12-18-secure-spring-rest-api.md (hunk 1 above)
Comment (bdemers) [reply _source/_posts/2018-12-18-secure-spring-rest-api.md]: they should get pulled automatically

File: _source/_posts/2018-12-18-secure-spring-rest-api.md (hunk 3 above)
Comment (bdemers) [reply _source/_posts/2018-12-18-secure-spring-rest-api.md]: Got a branch? i can take a look

Comment (mraible) [reply _source/_posts/2018-12-18-secure-spring-rest-api.md]: Nope, I started modifying https://github.com/oktadeveloper/okta-secure-spring-rest-api-example, but gave up once it didn't work.

Comment (raphaeldovale) [reply _source/_posts/2018-12-18-secure-spring-rest-api.md]: @mraible  and @bdemers I solved in another article:
https://github.com/raphaeldovale/use_postgresql_with_spring_boot/blob/security/src/main/java/com/okta/developer/springbootpostgressql/SpringBootPostgressqlApplication.java

Had to add more spring security code:
```java
        protected void configure(HttpSecurity http) throws Exception {
            http
                    .authorizeRequests().anyRequest().authenticated()
                    .and()
                    .oauth2ResourceServer().jwt();
}
```

**Pull request merged**